portal/src/util/passwordGenerator.ts

```diff
@@ -19,96 +19,93 @@ export interface RandSource {
   shuffle(list: string): string;
 }
 
-class CryptoRandSource implements RandSource {
-  randomBytes(n: number): Uint8Array {
+const cryptoRandSource: RandSource = {
+  randomBytes: (n: number): Uint8Array => {
     const array = new Uint8Array(n);
     window.crypto.getRandomValues(array);
     return array;
-  }
+  },
+  shuffle: (list: string): string => {
+    const array = list.split("");
+    for (let i = array.length - 1; i > 0; i--) {
+      const randomArray = new Uint8Array(1);
+      window.crypto.getRandomValues(randomArray);
+      const j = randomArray[0] % (i + 1);
+      [array[i], array[j]] = [array[j], array[i]];
+    }
+    return array.join("");
+  },
+};
```

Comment:
```
interface Source {
  // Return [0, n). Assume n <= 256
  intn(n: number): number;
}
```

This is Intn in JavaScript with Math.random https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random#getting_a_random_integer_between_two_values

```
const MAX_SAFE_INTEGER = Math.pow(2, 32) - 1;

if (n > MAX_SAFE_INTEGER) {
  throw new Error();
}
const numberBytesNeeded = Math.ceil(Math.log2(n) / 8);
forever {
  const array = new Uint8Array(numberBytesNeeded);
  window.crypto.getRandomValues(array);
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView/getUint32
  // Handle modular bias
}
```

Reply:
Updated.